### PR TITLE
Fix null-ref for MSAL WASM Auth

### LINKED ISSUE #30538: NullReferenceException thrown when making a call to secured API while not authenticated in Blazor WASM app with MSAL auth
### Describe the bug
I run a Blazor WASM app that uses MSAL for Azure AAD B2C auth, and can call into my ASP.net web api. Whenever the user is not authenticated and Blazor app tries to make a call to secured api via configured httpClient it results in a NullReferenceException being thrown, instead of expected AccessTokenNotAvailableException.
StackTrace:
```
         at Microsoft.AspNetCore.Components.WebAssembly.Authentication.AuthorizationMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)

```

I have narrowed it down to this bit of code:
 ```
 var tokenResult = _tokenOptions != null ?
      await _provider.RequestAccessToken(_tokenOptions) :
      await _provider.RequestAccessToken();

  if (tokenResult.TryGetToken(out var token))
  {
      _lastToken = token;
      _cachedHeader = new AuthenticationHeaderValue("Bearer", _lastToken.Value);
  }
  else
  {
      throw new AccessTokenNotAvailableException(_navigation, tokenResult, _tokenOptions?.Scopes);
  }
```

Basically when the user is not authenticated the tokenResult I get from provider will for whatever reason still have a success status, yet the inner token in that result is null. TryGetToken method on AccessTokenResult will only check for success status and will return true even if the token is null:

```
  public bool TryGetToken(out AccessToken accessToken)
  {
      if (Status == AccessTokenResultStatus.Success)
      {
          accessToken = _token;
          return true;
      }
      else
      {
          accessToken = null;
          return false;
      }
  }
```

So naturally, when it proceeds in the 1st bit of code it will pass the if statement and try to get a value from a null token, resulting in a NullReferenceException:

`_cachedHeader = new AuthenticationHeaderValue("Bearer", _lastToken.Value); // _lastToken is null here`

I may be wrong and it is an expected behavior, but according to documentation making a call to a secure API should result in a AccessTokenNotAvailableException instead of NullReferenceException.
https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/additional-scenarios?view=aspnetcore-5.0#attach-tokens-to-outgoing-requests

### To Reproduce

- Create a project via dotnet command described here: https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-5.0
`dotnet new blazorwasm -au IndividualB2C --aad-b2c-instance "{AAD B2C INSTANCE}" --api-client-id "{SERVER API APP CLIENT ID}" --app-id-uri "{SERVER API APP ID URI}" --client-id "{CLIENT APP CLIENT ID}" --default-scope "{DEFAULT SCOPE}" --domain "{TENANT DOMAIN}" -ho -o {APP NAME} -ssp "{SIGN UP OR SIGN IN POLICY}"`

- In Client/Pages/FetchData.razor remove "[Authorize]" attribute
- Run project
- Navigate to fetch data while being unauthenticated
- Unhandled NullReferenceException exception is thrown

### Exceptions (if any)
NullReferenceException

### Further technical details
- Microsoft.AspNetCore.Components.WebAssembly 5.0.3
- Microsoft.Authentication.WebAssembly.Msal 5.0.3

### LINKED ISSUE #32646: API call returns an Object reference not set to an instance of an object.
### Bug
```
crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Object reference not set to an instance of an object.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.Components.WebAssembly.Authentication.AuthorizationMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at System.Net.Http.Json.HttpClientJsonExtensions.
```

### Steps To Reproduce
Create a new Blazor Web Assembly .NET 5 project (ASP.NET Core hosted) with the Microsoft Identity platform, using Azure AD.

Add a new API Controller and add a Component to call the API. When calling the API you get the above error message. The ```Object reference not set to an instance of an object``` was a confusing one. If it's a lack of authentication, I think it should return a 401.  I think the above error has something to do with the HttpClient declared in ```Program.cs```:

```

            builder.Services.AddHttpClient("WebApplication1.ServerAPI", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();
```

But the API Controller does not have any authentication.

Simple repo at: https://github.com/mikecasas/webapplication1

### Further technical details
- Windows 10 Pro 20H2 Build 19042.985
- ASP.NET Core version ```5.0.23```

## Dotnet info

h:\>dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.203
 Commit:    383637d63f

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.203\

Host (useful for support):
  Version: 5.0.6
  Commit:  478b2f8c0e

.NET SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.400 [C:\Program Files\dotnet\sdk]
  3.1.409 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.20601.7 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]
  5.0.203 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

### PR DESCRIPTION
**Pull request by TanayParikh**
Currently, `TryGetToken` still returns success as long as `AccessTokenResultStatus.Success`, even if the `_token` itself is `null`:
https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AccessTokenResult.cs#L41-L53

This results in a `NullReferenceException` here:
https://github.com/dotnet/aspnetcore/blob/d7acbfb3c4d14f38eaaa213410e3c71752bdad9f/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AuthorizationMessageHandler.cs#L62

Now, an exception is thrown when an `account` isn't available. Consequently, we return `AccessTokenResultStatus.RequiresRedirect`. This goes into:

https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AccessTokenResult.cs#L50-L51

Resulting in:
https://github.com/dotnet/aspnetcore/blob/d7acbfb3c4d14f38eaaa213410e3c71752bdad9f/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AuthorizationMessageHandler.cs#L66

Verified this is working as expected locally. 

Fixes: https://github.com/dotnet/aspnetcore/issues/30538
Fixes: https://github.com/dotnet/aspnetcore/issues/32646

### CODE REVIEW

**Comment by TanayParikh** [comment]
Marking as draft as I'm still validating this solution.

**Comment by javiercn** [comment]
> Instead, of returning `true` from `TryGetToken` when `AccessTokenResultStatus.Success` despite a `null` token, we should return false, which would result in us throwing `AccessTokenNotAvailableException`.

If we have a null token, what's the reason for it, shouldn't we instead return a different status code instead?

**Comment by TanayParikh** [comment]
> If we have a null token, what's the reason for it

In this case, the user is unauthenticated. Repro steps:

>> In Client/Pages/FetchData.razor remove "[Authorize]" attribute
>> Run project
>> Navigate to fetch data while being unauthenticated
>> Unhandled NullReferenceException exception is thrown

---

> shouldn't we instead return a different status result instead?

I believe `AccessTokenNotAvailableException` should be appropriate based on the documentation here:

>> If a token can't be acquired, an AccessTokenNotAvailableException is thrown.
- https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/additional-scenarios?view=aspnetcore-6.0#attach-tokens-to-outgoing-requests

Was there an alternate exception you'd suggest?

**Comment by javiercn** [comment]
> Was there an alternate exception you'd suggest?

The end result should be the same, what I'm suggesting is that this might be a bug in AuthenticationService.ts instead

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts [hunk 1]
```diff
@@ -125,7 +125,7 @@ class MsalAuthorizeService implements AuthorizeService {
     async getTokenCore(scopes?: string[]): Promise<AccessToken | undefined> {
         const account = this.getAccount();
         if (!account) {
-            return;
+            throw new Error('Failed to retrieve token, no account found.');
```
Comment (TanayParikh) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts:128]: This exception is caught on line 119 above in the caller function. By throwing/catching the exception we no longer return a `AccessTokenResultStatus.Success` but rather `AccessTokenResultStatus.RequiresRedirect`. This goes into:

https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AccessTokenResult.cs#L50-L51

Resulting in:
https://github.com/dotnet/aspnetcore/blob/d7acbfb3c4d14f38eaaa213410e3c71752bdad9f/src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AuthorizationMessageHandler.cs#L66

Verified this is working as expected locally.

**Review by javiercn — APPROVED** [review]

**Comment by JCPuerto** [comment]
Hi @javiercn,

Do you know when this change will be available by any chance?

Thank you!

**Comment by TanayParikh** [comment]
> Do you know when this change will be available by any chance?

This'll be part of .NET 6.0 RC2 (mid October release).

**Comment by JCPuerto** [comment]
I'm using .NET 5. So do I have to wait until version 6? Is there a workaround in the meantime? I'm new to Blazor and I'd like to use this:

```
  catch (AccessTokenNotAvailableException exception)
  {
      exception.Redirect();
  }
```

Right now I'm just getting a plain Exception.

Thanks again!

**Comment by TanayParikh** [comment]
> I'm using .NET 5. So do I have to wait until version 6?

Yes, we don't currently plan on backporting this to .NET 5. 

If this is not for production purposes, you can checkout the .NET nightly build (in a day or two) which'll contain the fix:
https://github.com/dotnet/aspnetcore/blob/main/docs/DailyBuilds.md

**Pull request merged**